### DISCUSSION TITLE
Enable saving BMPs with a bit depth of 1

### DISCUSSION THREAD

**Discussion by thomasbarton** [libvips/libvips#2758] [Ideas]
My goal is to be able to convert an image to 1 bit depth and save it as a bmp.

I am able to convert it to having a 1 bit color depth with the function `vips_scRGB2BW` but when I go to save it with `vips_magicksave_buffer` there is currently no way to pass an image depth so the file is always an 8 bit file.  On the command line with imagemagick or graphicsmagick I am able to pass `-colors 2 +dither -type bilevel` to get the correct file format.

Any help or pointers in implementing this would be appreciated.

**Comment by jcupitt** [comment]
Hi @thomasbarton,

You'd need to dig about in the imagemagick core API and find the depth parameter, then add that to our imagemagick portability layer in `magick.c` (this has ifdefs for im6, im7 and gm). You'd need to add a `bitdepth` parameter to `magicksave` (this is the name we use for the parameter in the other savers), and then connect the two things up in the `magicksave_build` function.

For quantisation and dithering, you can either use the libvips one or add controls for the imagemagick one. The libvips one is usually libimagequant and is generally faster and better quality, so I'd be tempted to use that. Have a look at eg. `spngsave` or `gifsave` to see how to drive it.

We have two png savers -- `spngsave` is cleaner and simpler and does everything you need, so that would be a good place to start reading:

https://github.com/libvips/libvips/blob/master/libvips/foreign/spngsave.c

Good luck! I'd be happy to help with any followup questions.

**Reply by thomasbarton** [reply]
Thank you for the pointers!  When questions come up I will bring them back here.

**Comment by thomasbarton** [comment]
I am working with my coworker on this and one problem we ran into is that imagemagick does not seem to support bitdepth as an integer.  They have a set of enums to express bitdepth.

```
> image_type: the image type:   UndefinedType, BilevelType, GrayscaleType,
 %      GrayscaleAlphaType, PaletteType, PaletteAlphaType, TrueColorType,
 %      TrueColorAlphaType, ColorSeparationType, ColorSeparationAlphaType,
 %      or OptimizeType
```
ref https://imagemagick.org/api/MagickWand/magick-image_8c_source.html

I don't think that libvips would want to have multiple ways to express bitdepth, integer and imagemagick enums.  So we are thinking of exposing a boolean argument that is specific to bmps, `bmp-bitdepth-1`.  This is similar to how libvips currently has gif specific arguments for `optimize_gif_frames` in the same magicksave function.

What are your thoughts on this approach?

This is the PR that we are working on https://github.com/thomasbarton/libvips/pull/2

**Reply by jcupitt** [reply]
Hi, I think there is a depth setting, for example:

```
$ convert k2.jpg -depth 1 x.png
```

Writes a 4-bit palette PNG. I've not looked into the imagemagick API to see how to connect up to this setting though. They often do it via member variables, so I think I'd look there first.
